pilot/pkg/networking/plugin/authz/rbac.go

```diff
@@ -0,0 +1,393 @@
+func (Plugin) OnOutboundRouteConfiguration(in *plugin.InputParams, route *xdsapi.RouteConfiguration) {
+}
+
+// OnInboundRouteConfiguration implements the Plugin interface method.
+func (Plugin) OnInboundRouteConfiguration(in *plugin.InputParams, route *xdsapi.RouteConfiguration) {
+}
+
+// OnOutboundCluster implements the Plugin interface method.
+func (Plugin) OnOutboundCluster(env model.Environment, node model.Proxy, service *model.Service,
+	servicePort *model.Port, cluster *xdsapi.Cluster) {
+}
+
+func isRbacEnabled(store model.IstioConfigStore) bool {
+	var configProto *rbacproto.RbacConfig
+	config := store.RbacConfig(RbacConfigName, kube.IstioNamespace)
+	if config != nil {
+		configProto = config.Spec.(*rbacproto.RbacConfig)
+	}
+	if configProto == nil {
+		log.Debugf("rbac plugin disabled: No RbacConfig")
+		return false
+	}
+	// TODO(yangminzhu): Supports ON_WITH_INCLUSION and ON_WITH_EXCLUSION.
+	if configProto.Mode == rbacproto.RbacConfig_ON {
+		return true
+	}
+
+	log.Debugf("rbac plugin disabled by RbacConfig: %v", *configProto)
+	return false
+}
+
+// buildHTTPFilter builds the RBAC http filter that enforces the access control to the specified
+// service which is co-located with the sidecar proxy.
+func buildHTTPFilter(hostName model.Hostname, store model.IstioConfigStore) *http_conn.HttpFilter {
+	namespace := hostName.Namespace()
+	roles := store.ServiceRoles(namespace)
+	if roles == nil {
+		log.Debugf("no service role found for %s in namespace %s", hostName, namespace)
+	}
+	bindings := store.ServiceRoleBindings(namespace)
+	if bindings == nil {
+		log.Debugf("no service role binding found for %s in namespace %s", hostName, namespace)
+	}
+	config := convertRbacRulesToFilterConfig(hostName.String(), roles, bindings)
+	return &http_conn.HttpFilter{
+		Name:   RbacFilterName,
+		Config: util.MessageToStruct(config),
+	}
+}
+
+// convertRbacRulesToFilterConfig converts the current RBAC rules (ServiceRole and ServiceRoleBindings)
+// in service mesh to the corresponding proxy config for the specified service. The generated proxy config
+// will be consumed by envoy RBAC filter to enforce access control on the specified service.
+func convertRbacRulesToFilterConfig(service string, roles []model.Config, bindings []model.Config) *rbacconfig.RBAC {
+	// roleToBinding maps ServiceRole name to a list of ServiceRoleBindings.
+	roleToBinding := map[string][]*rbacproto.ServiceRoleBinding{}
+	for _, binding := range bindings {
+		bindingProto := binding.Spec.(*rbacproto.ServiceRoleBinding)
+		roleName := bindingProto.RoleRef.Name
+		if roleName == "" {
+			log.Errorf("ignored invalid binding with empty RoleRef.Name: %v", *bindingProto)
+			continue
+		}
+		roleToBinding[roleName] = append(roleToBinding[roleName], bindingProto)
+	}
+
+	rbac := &policyproto.RBAC{
+		Action:   policyproto.RBAC_ALLOW,
+		Policies: map[string]*policyproto.Policy{},
+	}
+
+start:
+	for _, role := range roles {
+		// Constructs the policy for each ServiceRole.
```

Comment:
I know that I had a comment about checking role eligibility first before convertPrincipals(). But it seems that after the change, the code readability is not that great, especially having to use "continue start" makes me a little uncomfortable. 

How about having the following logic? (I know it may go back to your previous code logic...)

```
for each role
   principals := convertToPrincipals(roleToBinding[role.Name])
   if len(principals) == 0 { 
      continue; // move to next role
   }
   declare permissions list
   for each rule inside the role {
       if the rule applies to the service {
          // convertToPermission and append to the permission list
       }
   }
   if permissions list is not empty {
      create a policy by combining principals and permissions
   }
}
```

Reply:
Done.